plugins/autoslab.cpp

```diff
@@ -0,0 +1,267 @@
+
+static command_result autoslab(color_ostream &out, std::vector<std::string> &parameters);
+
+DFHACK_PLUGIN("autoslab");
+DFHACK_PLUGIN_IS_ENABLED(is_enabled);
+
+REQUIRE_GLOBAL(world);
+
+// logging levels can be dynamically controlled with the `debugfilter` command.
+namespace DFHack
+{
+    // for configuration-related logging
+    DBG_DECLARE(autoslab, status, DebugCategory::LINFO);
+    // for logging during the periodic scan
+    DBG_DECLARE(autoslab, cycle, DebugCategory::LINFO);
+}
+
+static const auto CONFIG_KEY = std::string(plugin_name) + "/config";
+static PersistentDataItem config;
+enum ConfigValues
+{
+    CONFIG_IS_ENABLED = 0,
+    CONFIG_CYCLE_TICKS = 1,
+};
+static int get_config_val(int index)
+{
+    if (!config.isValid())
+        return -1;
+    return config.ival(index);
+}
+static bool get_config_bool(int index)
+{
+    return get_config_val(index) == 1;
+}
+static void set_config_val(int index, int value)
+{
+    if (config.isValid())
+        config.ival(index) = value;
+}
+static void set_config_bool(int index, bool value)
+{
+    set_config_val(index, value ? 1 : 0);
+}
+
+static int32_t cycle_timestamp = 0; // world->frame_counter at last cycle
+
+static command_result do_command(color_ostream &out, std::vector<std::string> &parameters);
+static void do_cycle(color_ostream &out);
+
+DFhackCExport command_result plugin_init(color_ostream &out, std::vector<PluginCommand> &commands)
+{
+    DEBUG(status, out).print("initializing %s\n", plugin_name);
+
+    // provide a configuration interface for the plugin
+    commands.push_back(PluginCommand(
+        plugin_name,
+        "Automatically engrave slabs of ghostly citizens!",
+        do_command));
+
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_enable(color_ostream &out, bool enable)
+{
+    if (!Core::getInstance().isWorldLoaded())
+    {
+        out.printerr("Cannot enable %s without a loaded world.\n", plugin_name);
+        return CR_FAILURE;
+    }
+
+    if (enable != is_enabled)
+    {
+        is_enabled = enable;
+        DEBUG(status, out).print("%s from the API; persisting\n", is_enabled ? "enabled" : "disabled");
+        set_config_bool(CONFIG_IS_ENABLED, is_enabled);
+    }
+    else
+    {
+        DEBUG(status, out).print("%s from the API, but already %s; no action\n", is_enabled ? "enabled" : "disabled", is_enabled ? "enabled" : "disabled");
+    }
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_shutdown(color_ostream &out)
+{
+    DEBUG(status, out).print("shutting down %s\n", plugin_name);
+
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_load_data(color_ostream &out)
+{
+    config = World::GetPersistentData(CONFIG_KEY);
+
+    if (!config.isValid())
+    {
+        DEBUG(status, out).print("no config found in this save; initializing\n");
+        config = World::AddPersistentData(CONFIG_KEY);
+        set_config_bool(CONFIG_IS_ENABLED, is_enabled);
+        set_config_val(CONFIG_CYCLE_TICKS, 1200);
```

Comment:
it would be fine to take this out and just use a constant value (e.g. see autochop.cpp). I've been using constants for the cycle period for most new/updated plugins and should probably take this example out of the sample skeleton code.